fog/ledger/server/src/router_handlers.rs

```diff
@@ -0,0 +1,345 @@
+        router_server_err_to_rpc_status("Auth: e client accept", err.into(), logger)
+    })?;
+
+    let mut response = LedgerResponse::new();
+    response.mut_auth().set_data(client_auth_response.into());
+    Ok(response)
+}
+
+#[allow(unused_variables)]
+/// Handles a client's query request.
+async fn handle_query_request<E>(
+    query: attest::Message,
+    enclave: E,
+    shard_clients: Vec<Arc<KeyImageStoreApiClient>>,
+    logger: Logger,
+) -> Result<LedgerResponse, RpcStatus>
+where
+    E: LedgerEnclaveProxy,
+{
+    let mut query_responses: BTreeMap<ResponderId, EnclaveMessage<NonceSession>> = BTreeMap::new();
+    let mut shard_clients = shard_clients.clone();
+    let sealed_query = enclave
+        .decrypt_and_seal_query(query.into())
+        .map_err(|err| {
+            router_server_err_to_rpc_status(
+                "Query: internal encryption error",
+                err.into(),
+                logger.clone(),
+            )
+        })?;
+
+    // The retry logic here is:
+    // Set retries remaining to RETRY_COUNT
+    // Send query and process responses
+    // If there's a response from every shard, we're done
+    // If there's a new store, repeat
+    // If there's no new store and we don't have enough responses, decrement
+    // RETRY_COUNT and loop
+    let mut remaining_retries = RETRY_COUNT;
+    while remaining_retries > 0 {
+        let multi_ledger_store_query_request = enclave
+            .create_multi_key_image_store_query_data(sealed_query.clone())
+            .map_err(|err| {
+                router_server_err_to_rpc_status(
+                    "Query: internal encryption error",
+                    err.into(),
+                    logger.clone(),
+                )
+            })?
+            .into();
+        let clients_and_responses =
+            route_query(&multi_ledger_store_query_request, shard_clients.clone())
+                .await
+                .map_err(|err| {
+                    router_server_err_to_rpc_status(
+                        "Query: internal query routing error",
+                        err,
+                        logger.clone(),
+                    )
+                })?;
+
+        let processed_shard_response_data = process_shard_responses(clients_and_responses)
+            .map_err(|err| {
+                router_server_err_to_rpc_status(
+                    "Query: internal query response processing",
+                    err,
+                    logger.clone(),
+                )
+            })?;
+
+        for (store_responder_id, new_query_response) in processed_shard_response_data
+            .new_query_responses
+            .into_iter()
+        {
+            query_responses.insert(store_responder_id, new_query_response.into());
+        }
+
+        if query_responses.len() >= shard_clients.len() {
+            break;
+        }
+
+        shard_clients = processed_shard_response_data.shard_clients_for_retry;
```

Comment:
The intent might have been to merge these lists (and then deduplicate somehow) rather than reassign it, which makes more sense because this is potentially pulling in connections for new shards we did not have connections with previously. So, maybe it should be `shard_clients.append(processed_shard_response_data.shard_clients_for_retry.clone())` but I'm not sure what the deduplication would look like there.

Reply:
Nick, you're absolutely right, that termination condition is wrong. I've "fixed" stuff related to that `shard_clients` reassignment like four times now and every time I do it's still wrong.

The termination logic here _in general_ is not really correct, because (as I mentioned in the PR description) it's assuming that a response from each shard is both necessary and sufficient to call a query finished. Neither of those is necessarily the case however - we could have shards configured with overlapping ranges (in which case we don't need to hear from all of them), or we could have shards backed by stores which are mis-configured to a different range than the range the shard is expected to cover (in which case we could get a response from every shard and still not have the full range covered). That all needs a pretty significant reworking of the shard management API to report responsible block ranges, and @samdealy wanted to defer talking about that until after the holiday (for good reasons).

I can't just use the logic from `router_request_handler.rs` in Fog View Router, because that logic may not work correctly in the case that `>RETRY_COUNT` new stores are added to a shard between queries - any case where a query reaches a store that needs authentication is treated as a "failure" for purposes of the `0..RETRY_COUNT` loop termination condition.

For the purposes of this PR, I am going to try to fix the issue Nick brought up and write a correct implementation of the "exactly one response from each shard" loop termination condition. This will probably involve not reassigning `shard_clients` inside the loop, because that keeps messing me up.